### PR TITLE
Overload permutedims for Diagonal

### PR DESCRIPTION
**Pull request by dlfivefifty**
This overloads `permutedims(D::Diagonal)` to just return `D`.

### CODE REVIEW

**Comment by mcabbott** [comment]
Should it copy? 

Similarly, `permutedims(::Transpose)` at present does make a copy, it could just un-wrap but I'd be concerned this would break expectations.

**Comment by dlfivefifty** [comment]
No need for copy:
```julia
julia> a = randn(5);

julia> p = permutedims(a);

julia> p[2] = 2
2

julia> a
5-element Array{Float64,1}:
  1.5407665908706154
  2.0
 -1.5463134371009768
  1.8914431479285454
 -0.24465440229260277
```

**Comment by mcabbott** [comment]
Oh I did not know that example. Count me as surprised!

**Comment by dlfivefifty** [comment]
Though note it's inconsistent:
```julia
julia> a = randn(5,1);

julia> p = permutedims(a);

julia> p[2] = 2;

julia> a
5×1 Array{Float64,2}:
  0.6280227134696655
  0.7033079386410532
 -0.07016519373621317
 -0.5413746174386154
  0.6870988321650711
```
Not sure if this is intentional, and it should probably be documented.

A bigger issue is that this could count as breaking if a user does:
```julia
D = Diagonal(randn(5))
P = permutedims(D)
P[1,2] = 3
```
but I personal consider assuming something about mutability of the return type to be a bad design choice, and it's very unlikely this exists in any user code.

**Comment by mcabbott** [comment]
Yes, that was my concern. If you thought that `permutedim`'s contract was to return a fresh `Array`, then it ever being lazy could be surprising, and giving a structured matrix even more so. 

Now maybe that isn't the rule, but what is? If it's a non-recursive `transpose` and sometimes lazy, then why doesn't it return `PermutedDimsArray(::Matrix)`? If it's specially for LinearAlgebra types, should `permutedims(UpperTriangular(rand(3,3)))` do something? 

(For the vector case the docstring does say "Reshape vector v" but I guess I never read it.)

**Comment by dlfivefifty** [comment]
Since `permutedims` doesn't provide a contract I'm taking it to be "buyer beware". But also there never was a guarantee of no breaking changes in StdLib, so I think it's permitted as a 1.x change if it's felt worthwhile.

I agree with you about overloading for other types, and if agreed that such a change is appropriate for 1.x I can add the following changes:
```julia
permutedims(A::UpperTriangular) = LowerTriangular(permutedims(parent(A)))
permutedims(A::UnitUpperTriangular) = UnitLowerTriangular(permutedims(parent(A)))
...
permutedims(A::Tridiagonal) = Tridiagonal(A.du, A.d, A.dl)
permutedims(A::SymTridiagonal) = A
permutedims(A::Symmetric) = A
permutedims(A::Hermitian) = Hermitian(permutedims(parent(A)), A.uplo == 'L' ? 'U' : 'L')
```

File: stdlib/LinearAlgebra/src/diagonal.jl [hunk 1]
```diff
@@ -580,6 +580,7 @@ transpose(D::Diagonal{<:Number}) = D
 transpose(D::Diagonal) = Diagonal(transpose.(D.diag))
 adjoint(D::Diagonal{<:Number}) = conj(D)
 adjoint(D::Diagonal) = Diagonal(adjoint.(D.diag))
+Base.permutedims(D::Diagonal) = D
```
Comment (vtjnash) [inline stdlib/LinearAlgebra/src/diagonal.jl:583]:
```suggestion
Base.permutedims(D::Diagonal) = D
Base.permutedims(D::Diagonal, perm) = (Base.checkdims_perm(D, D, perm); D)
```
IIUC

**Review by vtjnash — APPROVED** [review]

**Comment by vtjnash** [comment]
I also agree with adding those others too. Ping when you're happy with it, and I'll merge

**Comment by dlfivefifty** [comment]
I believe it's good too go now, assuming the tests pass

**Comment by vtjnash** [comment]
My sleep-addled brain thought you said it was too good now to merge. But I've merged this now, so if you want the other methods too, you can tag me as a reviewer again

**Comment by dlfivefifty** [comment]
Ah yes sorry forgot to address the other types. I don't have time to do that at the moment, and by the time I do have time I will have already forgotten this PR ever happened... so it will likely wait til I need it for something (unless someone else has a go).

**Pull request merged**